airbyte-commons-protocol/src/main/java/io/airbyte/commons/protocol/migrations/AirbyteMessageMigration.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.commons.protocol.migrations;
+
+import io.airbyte.commons.version.AirbyteVersion;
+
+/**
+ * AirbyteProtocol message migration interface
+ *
+ * @param <Old> The Old AirbyteMessage type
+ * @param <New> The New AirbyteMessage type
+ */
+public interface AirbyteMessageMigration<Old, New> {
+
+  /**
+   * Downgrades a message to from the new version to the old version
+   *
+   * @param message: the message to downgrade
+   * @return the downgraded message
+   */
+  Old downgrade(final New message);
+
+  /**
+   * Upgrades a message from the old version to the new version
+   *
+   * @param message: the message to upgrade
+   * @return the upgrade message
+   */
+  New upgrade(final Old message);
+
+  /**
+   * The Old version, note that due to semver, the important piece of information is the Major.
+   */
+  AirbyteVersion getOldVersion();
```

Comment:
Should we add a default that will return "0.2.0" here in order to avoid having the migration implementer to know that the default base version is 0.2.0 instead of another version

Reply:
I don't think a default make sense here. Each migration should come with its specific versions.
I wonder if it makes sense to only deal with Major versions explicitly, but I think we can address this later as we plug those objects in the pipeline.